src/Grpc.AspNetCore.Server/InterceptorCollection.cs

```diff
@@ -0,0 +1,83 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using Grpc.Core.Interceptors;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Grpc.AspNetCore.Server
+{
+    /// <summary>
+    /// Represents the pipeline of interceptors to be invoked when processing a gRPC call.
+    /// </summary>
+    public class InterceptorCollection : IReadOnlyList<InterceptorRegistration>
+    {
+        private static readonly IEnumerator<InterceptorRegistration> EmptyEnumerator = Enumerable.Empty<InterceptorRegistration>().GetEnumerator();
+
+        private List<InterceptorRegistration> _store;
+
+        /// <inheritdoc />
+        public bool IsEmpty => _store == null;
+
+        /// <inheritdoc />
+        public int Count => _store?.Count ?? 0;
+
+        /// <inheritdoc />
+        public InterceptorRegistration this[int index] => _store?[index];
+
+        /// <summary>
+        /// Add an interceptor to the end of the pipeline.
+        /// </summary>
+        /// <typeparam name="TInterceptor">The interceptor type.</typeparam>
+        /// <param name="args">The list of arguments to pass to the interceptor constructor when creating an instance.</param>
+        public void Add<TInterceptor>(params object[] args) where TInterceptor : Interceptor
```

Comment:
Actually we shouldn't have this API. This was taken from Middlewares that does not implement IMiddleware: https://github.com/aspnet/AspNetCore/blob/master/src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs#L59-L110. This is used to construct instances of middleware types which uses DI to inject services in the Invoke/InvokeAsync() methods. We are opting for constructor injection like IMiddleware.